### PR TITLE
CObject::Render

### PR DESCRIPTION
**Pull request by majesticCoding**
yacht particles

### CODE REVIEW

File: src/objects/Object.cpp [hunk 1]
```diff
@@ -128,15 +130,129 @@ CObject::Teleport(CVector vecPos)
...
+	float green = (0.8f * CTimeCycle::GetDirectionalGreen() + CTimeCycle::GetAmbientGreen_Obj()) * 165.75f;
+	float blue = (0.8f * CTimeCycle::GetDirectionalBlue() + CTimeCycle::GetAmbientBlue_Obj()) * 165.75f;
+
+	red = clamp(red, 0.0f, 255.0f);
+	green = clamp(green, 0.0f, 255.0f);
+	blue = clamp(blue, 0.0f, 255.0f);
+
+	int alpha = CGeneral::GetRandomNumberInRange(29, 196);
+
+	RwRGBA color{ red, green, blue, alpha };
```
Comment (Sergeanur) [inline src/objects/Object.cpp:152]: `RwRGBA color = { red, green, blue, alpha };`

File: src/objects/Object.cpp [hunk 2]
```diff
@@ -128,15 +130,129 @@ CObject::Teleport(CVector vecPos)
...
+		float moveSpeedMagnitude = this->GetMoveSpeed().Magnitude();
+		if (moveSpeedMagnitude > 0.0f) {
+			float scaleMax = CModelInfo::GetModelInfo(this->GetModelIndex())->GetColModel()->boundingBox.max.y * 0.85f;
+
+			CVector dir = this->GetMoveSpeed() + 0.3f * this->GetRight() - 0.5f * this->GetForward();
+			dir.z += 0.05f * moveSpeedMagnitude;
+
+			float posX = scaleMax * this->GetForward().x + 2.25f * this->GetRight().x + this->GetPosition().x;
+			float posY = scaleMax * this->GetForward().y + 2.25f * this->GetRight().y + this->GetPosition().y;
+			float posZ = scaleMax * this->GetForward().z + 2.25f * this->GetRight().z + this->GetPosition().z;
```
Comment (Sergeanur) [inline src/objects/Object.cpp:164]: This looks like `CVector pos = scaleMax * GetForward() + 2.25f * GetRight() + GetPosition()`, same with the rest down below

**Review by Sergeanur — CHANGES_REQUESTED** [review]

File: src/objects/Object.cpp (hunk 1 above)
Comment (majesticCoding) [reply src/objects/Object.cpp]: fixed

File: src/objects/Object.cpp (hunk 2 above)
Comment (majesticCoding) [reply src/objects/Object.cpp]: fixed

File: src/objects/Object.cpp [hunk 3]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
 
+	float red = (0.8f * CTimeCycle::GetDirectionalRed() + CTimeCycle::GetAmbientRed_Obj()) * 165.75f;
+	float green = (0.8f * CTimeCycle::GetDirectionalGreen() + CTimeCycle::GetAmbientGreen_Obj()) * 165.75f;
+	float blue = (0.8f * CTimeCycle::GetDirectionalBlue() + CTimeCycle::GetAmbientBlue_Obj()) * 165.75f;
+
+	red = clamp(red, 0.0f, 255.0f);
+	green = clamp(green, 0.0f, 255.0f);
+	blue = clamp(blue, 0.0f, 255.0f);
+
+	int alpha = CGeneral::GetRandomNumberInRange(29, 196);
```
Comment (aap) [inline src/objects/Object.cpp:150]: CGeneral::GetRandomNumberInRange(196, 225);

File: src/objects/Object.cpp [hunk 4]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+
+	RwRGBA color = { red, green, blue, alpha };
+
+	if (this->GetModelIndex() == MI_YT_MAIN_BODY) {
+		float moveSpeedMagnitude = this->GetMoveSpeed().Magnitude();
+		if (moveSpeedMagnitude > 0.0f) {
+			float scaleMax = CModelInfo::GetModelInfo(this->GetModelIndex())->GetColModel()->boundingBox.max.y * 0.85f;
```
Comment (aap) [inline src/objects/Object.cpp:157]: `float scaleMax = GetColModel()->boundingBox.max.y * 0.85f;`

File: src/objects/Object.cpp [hunk 5]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+	red = clamp(red, 0.0f, 255.0f);
+	green = clamp(green, 0.0f, 255.0f);
+	blue = clamp(blue, 0.0f, 255.0f);
+
+	int alpha = CGeneral::GetRandomNumberInRange(29, 196);
+
+	RwRGBA color = { red, green, blue, alpha };
+
+	if (this->GetModelIndex() == MI_YT_MAIN_BODY) {
+		float moveSpeedMagnitude = this->GetMoveSpeed().Magnitude();
```
Comment (aap) [inline src/objects/Object.cpp:155]: they actually recalculate this all the time...but i guess we don't have to do that too

File: src/objects/Object.cpp [hunk 6]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+		if (moveSpeedMagnitude > 0.0f) {
+			float scaleMax = CModelInfo::GetModelInfo(this->GetModelIndex())->GetColModel()->boundingBox.max.y * 0.85f;
+
+			CVector dir = this->GetMoveSpeed() + 0.3f * this->GetRight() - 0.5f * this->GetForward();
+			dir.z += 0.05f * moveSpeedMagnitude;
+
+			CVector pos = scaleMax * this->GetForward() + 2.25f * this->GetRight() + this->GetPosition();
+
+			float pfOutLevel;
```
Comment (aap) [inline src/objects/Object.cpp:164]: bad name, this is not a pointer. maybe just `float waterLevel;`?

File: src/objects/Object.cpp [hunk 7]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+			CWaterLevel::GetWaterLevel(pos.x, pos.y, pos.z, &pfOutLevel, true);
+			pos.z = pfOutLevel + 0.75f;
+
+			CParticle::AddParticle(PARTICLE_BOAT_SPLASH, pos, dir, nil, 1.2f * moveSpeedMagnitude, color,
+				CGeneral::GetRandomNumberInRange(0.0f, 0.4f), CGeneral::GetRandomNumberInRange(0.0f, 45.0f), 0, 0);
+
+			float scaleMin = CModelInfo::GetModelInfo(this->GetModelIndex())->GetColModel()->boundingBox.min.y;
```
Comment (aap) [inline src/objects/Object.cpp:171]: see above

File: src/objects/Object.cpp [hunk 8]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+
+			CParticle::AddParticle(PARTICLE_BOAT_SPLASH, pos, dir, nil, 0.9f, color,
+				CGeneral::GetRandomNumberInRange(0.0f, 0.4f), CGeneral::GetRandomNumberInRange(0.0f, 45.0f), 0, 0);
+		}
+	}
+
+	if (this->GetModelIndex() == MI_YT_MAIN_BODY2) {
+		float moveSpeedMagnitude = this->GetMoveSpeed().Magnitude();
+		if (moveSpeedMagnitude > 0.0f) {
+			float scaleMax = CModelInfo::GetModelInfo(this->GetModelIndex())->GetColModel()->boundingBox.max.y * 0.85f;
```
Comment (aap) [inline src/objects/Object.cpp:205]: see above

File: src/objects/Object.cpp [hunk 9]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+			CVector pos = scaleMax * this->GetForward() - 2.25f * this->GetRight() + this->GetPosition();
+
+			float pfOutLevel;
+			CWaterLevel::GetWaterLevel(pos.x, pos.y, pos.z, &pfOutLevel, true);
+			pos.z = pfOutLevel + 0.75f;
+
+			CParticle::AddParticle(PARTICLE_BOAT_SPLASH, pos, dir, nil, 1.2f * moveSpeedMagnitude, color,
+				CGeneral::GetRandomNumberInRange(0.0f, 0.4f), CGeneral::GetRandomNumberInRange(0.0f, 45.0f), 0, 0);
+
+			float scaleMin = CModelInfo::GetModelInfo(this->GetModelIndex())->GetColModel()->boundingBox.min.y;
```
Comment (aap) [inline src/objects/Object.cpp:219]: see above

File: src/objects/Object.cpp [hunk 10]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+				CGeneral::GetRandomNumberInRange(0.0f, 0.4f), CGeneral::GetRandomNumberInRange(0.0f, 45.0f), 0, 0);
+		}
+	}
+
+	if (this->GetModelIndex() == MI_YT_MAIN_BODY2) {
+		float moveSpeedMagnitude = this->GetMoveSpeed().Magnitude();
+		if (moveSpeedMagnitude > 0.0f) {
+			float scaleMax = CModelInfo::GetModelInfo(this->GetModelIndex())->GetColModel()->boundingBox.max.y * 0.85f;
+
+			CVector dir = this->GetMoveSpeed() + 0.3f * this->GetRight() - 0.5f * this->GetForward();
```
Comment (aap) [inline src/objects/Object.cpp:207]: `- 0.3f ...`

File: src/objects/Object.cpp [hunk 11]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+			dir.z += 0.05f * moveSpeedMagnitude;
+
+			CVector pos = scaleMax * this->GetForward() - 2.25f * this->GetRight() + this->GetPosition();
+
+			float pfOutLevel;
```
Comment (aap) [inline src/objects/Object.cpp:212]: see above

File: src/objects/Object.cpp [hunk 12]
```diff
@@ -128,15 +130,115 @@ CObject::Teleport(CVector vecPos)
...
+	float green = (0.8f * CTimeCycle::GetDirectionalGreen() + CTimeCycle::GetAmbientGreen_Obj()) * 165.75f;
+	float blue = (0.8f * CTimeCycle::GetDirectionalBlue() + CTimeCycle::GetAmbientBlue_Obj()) * 165.75f;
+
+	red = clamp(red, 0.0f, 255.0f);
+	green = clamp(green, 0.0f, 255.0f);
+	blue = clamp(blue, 0.0f, 255.0f);
+
+	int alpha = CGeneral::GetRandomNumberInRange(196, 225);
+
+	RwRGBA color = { red, green, blue, alpha };
```
Comment (ShFil119) [inline src/objects/Object.cpp:152]: It prevents building on mac. Casting each arg to uint8 should be easiest way to solve this issue.

**Pull request merged**